File: backend/shared/repositories/firestore_job_repository.py
"""Firestore repository implementation."""

from datetime import datetime, timezone

from google.cloud import firestore

from backend.shared.clients.firestore import get_firestore_client
from backend.shared.models.jobs import JobRecord
from backend.shared.config import get_settings


class FirestoreJobRepository:
    """Firestore-backed repository for validation jobs."""

    def __init__(self) -> None:
        settings = get_settings()
        self._client = get_firestore_client()
        self._collection = self._client.collection(settings.firestore_collection)

    def save(self, job: JobRecord) -> JobRecord:
        self._collection.document(job.job_id).set(self._serialize(job))
        return job

    def get(self, job_id: str) -> JobRecord | None:
        snapshot = self._collection.document(job_id).get()
        if not snapshot.exists:
            return None
        data = snapshot.to_dict() or {}
        return self._deserialize(data)

    def update(self, job: JobRecord) -> JobRecord:
        self._collection.document(job.job_id).set(self._serialize(job))
        return job

    def list_page(self, page: int, page_size: int) -> tuple[list[JobRecord], bool]:
        offset = max(page - 1, 0) * page_size
        query = (
            self._collection.order_by("updated_at", direction=firestore.Query.DESCENDING)
            .offset(offset)
            .limit(page_size + 1)
        )
        records = [self._deserialize(snapshot.to_dict() or {}) for snapshot in query.stream()]
        has_next = len(records) > page_size
        return records[:page_size], has_next

    def _serialize(self, job: JobRecord) -> dict:
        data = job.model_dump(mode="json")
        return data

    def _deserialize(self, data: dict) -> JobRecord:
        if isinstance(data.get("created_at"), str):
            data["created_at"] = self._parse_datetime(data["created_at"])
        if isinstance(data.get("updated_at"), str):
            data["updated_at"] = self._parse_datetime(data["updated_at"])
        return JobRecord.model_validate(data)

    def _parse_datetime(self, value: str) -> datetime:
        return datetime.fromisoformat(value.replace("Z", "+00:00")).astimezone(
            timezone.utc
        )

File: packages/infrastructure/src/repositories/firestore-job-repository.ts
import { Firestore, Timestamp } from '@google-cloud/firestore';
import type { JobRecord } from '@domain';
import { getInfrastructureSettings } from '../config/settings';
import type { JobRepository } from './job-repository';

function parseDate(value: unknown): string | undefined {
  if (!value) {
    return undefined;
  }

  if (typeof value === 'string') {
    return value;
  }

  if (value instanceof Date) {
    return value.toISOString();
  }

  if (value instanceof Timestamp) {
    return value.toDate().toISOString();
  }

  if (
    typeof value === 'object' &&
    value !== null &&
    'toDate' in value &&
    typeof (value as { toDate: unknown }).toDate === 'function'
  ) {
    const timestampLike = value as { toDate: () => Date };
    return timestampLike.toDate().toISOString();
  }

  return undefined;
}

export class FirestoreJobRepository implements JobRepository {
  private readonly firestore: Firestore;
  private readonly collectionName: string;

  constructor(firestore?: Firestore) {
    const settings = getInfrastructureSettings();
    this.firestore =
      firestore ??
      new Firestore({
        projectId: settings.gcpProjectId,
        databaseId: settings.firestoreDatabase,
      });
    this.collectionName = settings.firestoreCollection;
  }

  async save(job: JobRecord): Promise<JobRecord> {
    await this.collection().doc(job.jobId).set(this.serialize(job));
    return job;
  }

  async get(jobId: string): Promise<JobRecord | null> {
    const snapshot = await this.collection().doc(jobId).get();
    if (!snapshot.exists) {
      return null;
    }

    return this.deserialize(snapshot.data() ?? {});
  }

  async update(job: JobRecord): Promise<JobRecord> {
    await this.collection().doc(job.jobId).set(this.serialize(job));
    return job;
  }

  async listPage(page: number, pageSize: number): Promise<{ records: JobRecord[]; hasNext: boolean }> {
    const offset = Math.max(page - 1, 0) * pageSize;
    const snapshots = await this.collection()
      .orderBy('updated_at', 'desc')
      .offset(offset)
      .limit(pageSize + 1)
      .get();

    const records = snapshots.docs.map((document) => this.deserialize(document.data() ?? {}));
    return {
      records: records.slice(0, pageSize),
      hasNext: records.length > pageSize,
    };
  }

  private collection() {
    return this.firestore.collection(this.collectionName);
  }

  private serialize(job: JobRecord): Record<string, unknown> {
    return {
      job_id: job.jobId,
      merchant_id: job.merchantId,
      request_id: job.requestId ?? null,
      status: job.status,
      poll_count: job.pollCount,
      request: job.request,
      progress: job.progress ?? null,
      overall_result: job.overallResult ?? null,
      documents: job.documents ?? null,
      normalized_snapshot: job.normalizedSnapshot ?? null,
      cross_validation: job.crossValidation ?? null,
      created_at: job.createdAt,
      updated_at: job.updatedAt,
    };
  }

  private deserialize(data: Record<string, unknown>): JobRecord {
    const jobId = data.job_id ?? data.jobId;
    const merchantId = data.merchant_id ?? data.merchantId;
    const requestId = data.request_id ?? data.requestId;
    const pollCount = data.poll_count ?? data.pollCount ?? 0;
    const overallResult = data.overall_result ?? data.overallResult ?? null;
    const normalizedSnapshot = data.normalized_snapshot ?? data.normalizedSnapshot ?? null;
    const crossValidation = data.cross_validation ?? data.crossValidation ?? null;
    const createdAt = data.created_at ?? data.createdAt;
    const updatedAt = data.updated_at ?? data.updatedAt;

    return {
      jobId: String(jobId),
      merchantId: String(merchantId),
      status: data.status as JobRecord['status'],
      pollCount: Number(pollCount),
      request: (data.request ?? {}) as Record<string, unknown>,
      requestId: (requestId as string | null | undefined) ?? null,
      progress: (data.progress as JobRecord['progress']) ?? null,
      overallResult: (overallResult as JobRecord['overallResult']) ?? null,
      documents: (data.documents as JobRecord['documents']) ?? null,
      normalizedSnapshot: (normalizedSnapshot as JobRecord['normalizedSnapshot']) ?? null,
      crossValidation: (crossValidation as JobRecord['crossValidation']) ?? null,
      createdAt: parseDate(createdAt) ?? new Date().toISOString(),
      updatedAt: parseDate(updatedAt) ?? new Date().toISOString(),
    };
  }
}
